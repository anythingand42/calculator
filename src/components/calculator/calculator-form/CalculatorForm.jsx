import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Input from '../../input';
import Select from '../../select';
import Button from '../../button';
import isNumeric from '../../../math-utils/isNumeric';
import { ActionEnum, operations } from '../shared/const';

const Container = styled.div`
  width: 100%;
  input, select, button {
    font-size: 16px;
  }
`;

const InputBox = styled.div`
  width: 100%;

  display: flex;
  margin-bottom: 0.5em;

  @media only screen and (min-width: 768px) {
    justify-content: space-between;
    & > select {
      margin-left: 1em;
      margin-right: 1em;
    }
  }

  @media only screen and (max-width: 767px) {
    flex-direction: column;
    & > *:not(:last-child) {
      margin-bottom: 0.5em;
    }
  }
`;

const ButtonBox = styled.div`
  display: flex;
  justify-content: flex-end;
`;

function CalculatorForm({
  A,
  B,
  operation,
  dispatch,
}) {
  return (
    <Container>
      <InputBox>
        <Input
          data-testid="input-a"
          type="text"
          spellcheck="false"
          placeholder="A"
          value={A}
          onChange={(event) => {
            event.preventDefault();
            dispatch({ type: ActionEnum.setA, payload: event.target.value });
          }}
          error={A && !isNumeric(A)}
        />
        <Select
          data-testid="operation-select"
          value={operation}
          onChange={(event) => {
            event.preventDefault();
            dispatch({ type: ActionEnum.setOperation, payload: event.target.value });
          }}
        >
          {Object.entries(operations).map(
            (([key, label]) => <option key={key} value={key}>{label}</option>),
          )}
        </Select>
        <Input
          data-testid="input-b"
          type="text"
          spellCheck={false}
          placeholder="B"
          value={B}
          onChange={(event) => {
            event.preventDefault();
            dispatch({ type: ActionEnum.setB, payload: event.target.value });
          }}
          error={B && !isNumeric(B)}
        />
      </InputBox>
      <ButtonBox>
        <Button
          data-testid="calculate"
          type="button"
          disabled={!(isNumeric(A) && isNumeric(B))}
          onClick={(event) => {
            event.preventDefault();
            dispatch({ type: ActionEnum.calculate });
          }}
        >
          Calculate
        </Button>
      </ButtonBox>
    </Container>
  );
}

CalculatorForm.propTypes = {
  A: PropTypes.string.isRequired,
  B: PropTypes.string.isRequired,
  operation: PropTypes.number.isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default CalculatorForm;
