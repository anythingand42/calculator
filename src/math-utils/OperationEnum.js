const OperationEnum = {
  sum: 0,
  divide: 1,
  mod: 2,
  maxPrimeBetween: 3,
};

export default OperationEnum;
