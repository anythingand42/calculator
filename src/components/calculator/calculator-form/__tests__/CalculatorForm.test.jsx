import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import CalculatorForm from '../CalculatorForm';
import OperationEnum from '../../../../math-utils/OperationEnum';

test('CalculatorForm component', () => {
  const { container } = render(<CalculatorForm A="2" B="4" operation={OperationEnum.sum} dispatch={() => {}} />);
  expect(container.firstChild).toMatchSnapshot();
});

test('CalculatorForm component in an invalid state', () => {
  const { container } = render(<CalculatorForm A="ababa" B="lalala" operation={OperationEnum.divide} dispatch={() => {}} />);
  expect(container.firstChild).toMatchSnapshot();
});
