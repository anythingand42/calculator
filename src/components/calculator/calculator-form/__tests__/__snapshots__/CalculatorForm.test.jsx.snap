// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorForm component 1`] = `
.c2 {
  border: 1px solid #ccc;
  outline: none;
  box-sizing: border-box;
  padding: 4px;
  min-height: 28px;
}

.c3 {
  border: 1px solid #ccc;
  outline: none;
  box-sizing: border-box;
  background: none;
  min-height: 28px;
}

.c5 {
  min-height: 28px;
}

.c0 {
  width: 100%;
}

.c0 input,
.c0 select,
.c0 button {
  font-size: 16px;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5em;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

@media only screen and (min-width:768px) {
  .c1 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }

  .c1 > select {
    margin-left: 1em;
    margin-right: 1em;
  }
}

@media only screen and (max-width:767px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .c1 > *:not(:last-child) {
    margin-bottom: 0.5em;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2"
      data-testid="input-a"
      placeholder="A"
      type="text"
      value="2"
    />
    <select
      class="c3"
      data-testid="operation-select"
    >
      <option
        value="0"
      >
        + (sum)
      </option>
      <option
        value="1"
      >
        / (divide)
      </option>
      <option
        value="2"
      >
        % (mod)
      </option>
      <option
        value="3"
      >
        max prime between
      </option>
    </select>
    <input
      class="c2"
      data-testid="input-b"
      placeholder="B"
      spellcheck="false"
      type="text"
      value="4"
    />
  </div>
  <div
    class="c4"
  >
    <button
      class="c5"
      data-testid="calculate"
      type="button"
    >
      Calculate
    </button>
  </div>
</div>
`;

exports[`CalculatorForm component in an invalid state 1`] = `
.c2 {
  border: 1px solid #ccc;
  outline: none;
  box-sizing: border-box;
  padding: 4px;
  min-height: 28px;
  border-color: red;
  color: red;
}

.c3 {
  border: 1px solid #ccc;
  outline: none;
  box-sizing: border-box;
  background: none;
  min-height: 28px;
}

.c5 {
  min-height: 28px;
}

.c0 {
  width: 100%;
}

.c0 input,
.c0 select,
.c0 button {
  font-size: 16px;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5em;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

@media only screen and (min-width:768px) {
  .c1 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }

  .c1 > select {
    margin-left: 1em;
    margin-right: 1em;
  }
}

@media only screen and (max-width:767px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .c1 > *:not(:last-child) {
    margin-bottom: 0.5em;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2"
      data-testid="input-a"
      placeholder="A"
      type="text"
      value="ababa"
    />
    <select
      class="c3"
      data-testid="operation-select"
    >
      <option
        value="0"
      >
        + (sum)
      </option>
      <option
        value="1"
      >
        / (divide)
      </option>
      <option
        value="2"
      >
        % (mod)
      </option>
      <option
        value="3"
      >
        max prime between
      </option>
    </select>
    <input
      class="c2"
      data-testid="input-b"
      placeholder="B"
      spellcheck="false"
      type="text"
      value="lalala"
    />
  </div>
  <div
    class="c4"
  >
    <button
      class="c5"
      data-testid="calculate"
      disabled=""
      type="button"
    >
      Calculate
    </button>
  </div>
</div>
`;
